### DISCUSSION TITLE
Is getServerSideProps becoming irrelevant?

### DISCUSSION THREAD

**Discussion by alex-r89** [vercel/next.js#15572] [Help] [answered]
Hi,

Just to clarify; I do understand how both of these methods work (ie build time vs request time rendering), however I find that `getStaticProps` seems to work well in a situation where I thought I would need to use `getServerSideProps`. Therefore, I am not sure If I am missing something.

To explain with context of my situation:

I am creating a web application with a paginated list of user uploaded images (ie `/page/${pageNumber}`). Users can upload, delete and vote on their favorite images. This lead me to think that I would need to use `getServerSideProps`, because I would like to see if users updates their photos (or gets a vote on their photo) more often than at build time. Saying that however, it seems to be that I can take from the `with-apollo` example and have queries run on the client (see [here](https://github.com/vercel/next.js/blob/canary/examples/with-apollo/components/PostList.js)) to update the data more often than just at build time.

This therefore leads me to believe that I don't need `getServerSideProps` at all? I could just use `getStaticProps`, `getStaticPaths` to build all of the paginated pages, and then have queries fire on the client to update this data? I also find that my page loads twice as fast if I use `getStaticProps` over `getServerSideProps` (Render starts at 1.5s with SSG vs almost 4s with SSR on Moto G4 3G network).

If this is correct, I am wondering why I would ever use `getServerSideProps`? Am I missing something?

Thanks

**Comment by leerob** [comment, marked as answer]
With improvements to `getStaticProps`/`getStaticPaths` and Incremental Static Regeneration, you can use static generation for maximum performance without sacrificing the benefits of server-side rendering. I wrote more about [this here](https://vercel.com/blog/nextjs-server-side-rendering-vs-static-generation), if you want to see some tangible examples.

**Reply by alex-r89** [reply]
Thanks very much @leerob .

To clarify then, it seems like my situation is completly solved by just using ` unstable_revalidate` (which I think I can use as `revalidate` in next 9.5).

If I was to have `revalidate` set to say 30 minutes, my issue of data going stale  and newly uploaded photos or votes on photos not showing up is solved - at least every 30 minutes?

**Reply by leerob** [reply]
Correct!

**Reply by alex-r89** [reply]
Hi @leerob 

I have been playing around with my code based on your blog post, it seems great so far but I have one follow up:

In the blog post, you mention how and Individual Product Page can use Static Generation with Data. You say how we can use fallback so that we dont need to generate 100,000 product pages at build time (I assume this means the ones we dont generate will have client side fetches).

What would happen though if we had 100 products initially, all statically generated - and then added 5 products after build. Would, for example, product 105 always be client side rendered? If product 105 just so happened to be the most popular product, would this page forever not be statically generated?

**Reply by alanqchen** [reply]
The way `fallback` works is when a page is fetched for the first time that hasn't been generated already, the fallback page will be displayed while the server generates the page. From there on after, the generated page will be served on every subsequent request as a statically generated page (and even with `revalidate`). So no, it would not keep using the fallback.

**Reply by alex-r89** [reply]
Thanks very much for clarifying @alanqchen  👍

**Comment by arso1er** [comment]
While `getStaticProps / getStaticPaths` are awesome, one use case I find `getServerSideProps` to be better is for search.

**Comment by saisandeepvaddi** [comment]
No, it's not becoming irrelevant. I generally use `getServerSideProps` when something needs to be checked on every request where page should be different for each user viewing the screen. Like if I have to fetch content only if user is logged in, I extract `req.session` in `getServerSideProps` and get data for that user and pass data in props accordingly. For private routes I'm not sure how `getStaticProps` can be alternative.

If there is no such requirement,  `getStaticProps` with `revalidate` works great. Blog posts are one such use case.

These two choices are what makes `NextJS` usage simple and great. It can be used for building huge admin dashboard applications as well as purely static websites.

**Reply by eric-burel** [reply]
Completely agree with this, that's what makes Next different from Gatsby for instance: you can use it for server-rendered private apps. For this you indeed need `getServerSideProps`.

**Reply by alex-r89** [reply]
But you can do this with `getStaticProps` and then use a client side fetch for the user data you need? This is how Vercel does their dashboard afaik.

**Reply by eric-burel** [reply]
Yeah but then what's the point of SSR? Many applications value comes from private, user-owned data, not from static data. Rendering a layout + a few static info is cool but often insufficient.

**Reply by martinbavio** [reply]
@eric-burel you might have misunderstood what @alex-r89 was trying to say. If you have data that needs to be private, you have two options while in Next.js land:

**a. SSR:** to be served-rendered on every request via `getServerSideProps`, as you perfectly explained.
**b. SPA:** To be client-rendered, which is the classic single page application technique: you load a page with an empty shell and then on the client, trigger a fetch request to load that private data only if necessary.

In a way, (a) was the reason why we were initially drawn into Next.js some years ago, since we were basically escaping all the craziness that was going on around (b). But I believe things have changed, and ironically enough, (b) can be superior because of two main reasons:

**1. Cache:** while it's possible to setup cache on `gSSP`, caching mechanisms in the client have gone a long way since the origins of the Jamstack. Libraries like `react-query` and `SWR` are simple, convenient and really smart when it comes to fetching data and managing the cache. I feel like cheating every time I use react-query.

**2. Preload:** _empty shell_ pages can actually be not as dumb as they look like. If you could anticipate what data a page might need, you could quite easily add a preload link using `<Head>` that would make remote data to be ready as soon as React/Next takes over. This is indeed unique to the server-rendered nature of Next, only that in a different way.

So as you can see, SPAs can actually be quite interesting to explore in a Next.js way, and since SEO is not relevant at all in this case, server-rendering should not be a requirement nor a rule.

In the end, the most interesting thing is that Next.js is making all think creatively about alternatives, based on its unique hybrid nature.

**Reply by eric-burel** [reply]
Yeah I am playing the devil's advocate, because I am still trying to figure out the best usage of dynamic SSR in private apps, as full dynamic rendering is very resource consuming. I am using Apollo GraphQL so I haven't had the chance to test `react-query` and `swr`, though they seem very promising.
So if I get it right, what you mean is that instead of trying to serve the data immediately, you focus more on preparing the data fetching step during the static render? So then you still have great perceived performance, without actually rendering dynamically?

Also, one difficulty that arises during static render and I can only figure out with dynamic render, is that since the page is private, by definition it can't be rendered by the server, since it's kind of a non-authenticated user. That makes parsing the React tree difficult. I guess you have to put all data requirements on top of the page, within the `getStaticProps`, you can't tie data fetching to the inner components?

**Reply by alevosia** [reply]
You just statically generate the dashboard UI but without the data then let the client fetch and fill in the data.

**Comment by jnv** [comment]
We ran into these reasons why we picked `getServerSideProps` even for public content which doesn't change often:

- You can't change HTTP status code when you use `getStaticProps`, per https://github.com/vercel/next.js/discussions/10960#discussioncomment-1201 (basically your only option is to do [Soft 404](https://support.google.com/webmasters/answer/181708) which is a bad practice).
- For the time being you don't have any option to force page regeneration other than wait for `revalidate`; e.g. you can't regenerate the page when you receive a webhook from CMS (but [it is planned](https://nextjs.org/blog/next-9-5#stable-incremental-static-regeneration)).
  - In our case we leverage CDN to cache the SSR page which we can flush directly.
  - Also a common pattern is to set `revalidate` really low and Next.js will regenerate the page if the output of `getStaticProps` changes, see: https://github.com/vercel/next.js/discussions/11552#discussioncomment-52146

**Reply by omar-dulaimi** [reply]
Hello @jnv 
Could you please elaborate on how you cache SSR pages?

**Reply by errnesto** [reply]
I also have a problem with the soft 404 and would love to hear If someone has a good solution for this :-)

**Reply by jnv** [reply]
The good news is that recently an experimental support for blocking fallback was merged in #15672. This is a prerequisite for controlling response status code so I hope there will be at least some plan soon-ish.

@omar-dulaimi Nothing special. We have Cloudfront in front of our app and we set appropriate `Cache-Control` header in the response. From memory we use the following directives: `public, max-age=0, s-maxage=900, must-revalidate`. This forces the browser to revalidate  HTML on every request (so it doesn't request stale assets), but keeps it cached on CDN for up to 15 minutes.
Unfortunately Cloudfront doesn't support `stale-while-revalidate` and generally it is pretty barebones compared to competitors.

**Reply by omar-dulaimi** [reply]
@jnv So if you have user login, would that be affected by caching?
like the user profile image for example

**Reply by jnv** [reply]
Well, in our case we don't have login so we didn't really have to consider it. Any kind of logic depending on user's state is handled client-side. Obviously you won't benefit much from CDN if you need authentication and want to use SSR.

The safe option is to opt-out of CDN caching completely and add `private` cache-control directive.
Some CDNs, including Cloudfront, can include cookie values in cache key, so you can cache server-side for each user individually based e.g. on their session cookie. That means each authenticated user will hit the origin server at least once for server-side rendered page.

**Reply by omar-dulaimi** [reply]
Awesome tip about unique cache per user. Any tutorial/resources to read more about this? I currently use a library called cacheable-response but it's not working fine for me since I have user login,etc.

**Reply by jnv** [reply]
I think `cacheable-response` can be easily adjusted to work for you, you can provide your own [getKey](https://github.com/Kikobeats/cacheable-response#getkey) function to include some information about user from request object. For example if you use a session middleware and you store user's ID in the session, you can do something like: 

```js
({req}) => `${req.session.user_id}_${normalizeUrl(req.url)}`
```

This is probably better than using CDN since session storage is something you can trust, unlike user input like cookie value.

As for CDN, here's [documentation for Cloudfront](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Cookies.html) – basically you can whitelist cookie containing user's session ID to be forwarded by CDN to origin. Note that this will probably create an attack vector since CDN is just a dumb key value store. In theory you could "guess" another user's cookie (even using a brute force attack) and receive content cached for them, since CDN won't validate any data associated with the session (e.g. IP address, user agent, validity of session). It would be probably possible to employ Lambda@Edge or something similar for that, but it really depends on your use case.

My personal opinion is that CDN caching for server-side rendered pages with authentication is not worth it. Especially if you are using framework like Next.js where you can sprinkle auth-dependent features (like user's profile image, logout link etc.) as client-side functionality on top of statically rendered pages.

**Comment by jaybytez** [comment]
I just wanted to add context for anyone searching this issue.  We actually are considering converting all our getStaticProps/getStaticPaths over to getServerSideProps because of the following:

1) With a CDN fronting the Next.js apps for us, Next.js only gets hit when the route is not cached or we have flushed the CDN cache.  Therefore, even with revalidate on SSG, you never see the revalidated content unless you flush the CDN cache.  So we keep the revalidate super low, then if a publisher pushes new content, a webhook flushes the CDN which allows the request to hit Next.js (of course, we should use fallback: 'unstable_blocking' to ensure the latest version is returned and cached).
2) Bigger issue is the fact that our whole website is not managed by Next.js, only a small portion of the large site and therefore SSG doesn't have access to the full url path (like dummy.com/blogs vs dummy.com/es/blogs vs dummy.com/in/blogs) - so if blogs is our application/baseURL, we don't have access to any portion of the URL prior to /blogs like the siteid/locale.
3) Related issue with localization/internationalization is even if we had the ability to access the full URL, that if we hit a page like dummy.com/blogs/post1 and dummy.com/es/blogs/post1 - SSG sees these two requests from the context path as /blogs/post1.  But we essentially have to build two versions of the page because one is for English (default) and one is for Spanish (along with the locale specific meta-data).  So we can't do this because we don't have the full URL to make the route unique per locale/site (and query parameters - if accessible in SSG - don't make a route unique either).  So the route for /blogs/post1 is the same to Next.js regardless.

So SSR for us would solve the ability to have access to the request in the Page through getServerSideProps (wish getStaticProps with fallback=true had access to the request).  And it would also solve generating a unique request every time and then our CDN could cache the full context URL against the dynamically created page for the specific language.  

I might be missing something here.
